plugins/common/parsing/parsing.go

```diff
@@ -0,0 +1,203 @@
+package parsing
+
+import (
+	"errors"
+	"fmt"
+	"github.com/influxdata/telegraf"
+	"strconv"
+	"strings"
+)
+
+var (
+	ErrEmptyDelimiter = errors.New("parsing: delimiter must not be empty")
+	ErrEmptyWildcard  = errors.New("parsing: wildcard must not be empty")
+)
+
+type Config struct {
+	config    []ConfigEntry
+	baseName  string
+	delimiter string
+	wildcard  string
+	log       telegraf.Logger
+}
+
+// NewConfig creates a new parser to extract measurements, tags and fields from structures like
+// mqtt topic or nats subjects.
+// each entry in the config describes how this structure should be interpreted.
+// baseName is just the individual name of the structure (e.g: 'topic' for mqtt).
+// delimiter is the character or string that divides the topic in different parts.
+// wildcard is the character or string that tells that a part of the topic can be any string.
+func NewConfig(entries []ConfigEntry,
+	baseName string,
+	delimiter string,
+	wildcard string,
+	log telegraf.Logger) *Config {
+	return &Config{
+		config:    entries,
+		baseName:  baseName,
+		delimiter: delimiter,
+		wildcard:  wildcard,
+		log:       log,
+	}
+}
+
+// ConfigEntry describes how a specific structure can be parsed into a telegraf.Metric.
+// And incoming structure is first matched against the Base string. This string can include wildcards.
+// If the incoming structure matches, measurement name, tags and fields can be extracted from it
+// and are added to the metric.
+type ConfigEntry struct {
+	Base        string            `toml:"topic"`
+	Measurement string            `toml:"measurement"`
+	Tags        string            `toml:"tags"`
+	Fields      string            `toml:"fields"`
+	FieldTypes  map[string]string `toml:"types"`
+
+	measurementIndex int
+	splitTags        []string
+	splitFields      []string
+	splitSubject     []string
+}
+
+// Init validates and configures the parser and has to be called before using it.
+func (c *Config) Init() error {
+	if c.wildcard == "" {
+		return ErrEmptyWildcard
+	}
+
+	if c.baseName == "" {
+		c.baseName = "topic"
+	}
+
+	if c.delimiter == "" {
+		return ErrEmptyDelimiter
+	}
+
+	for i := range c.config {
+		splitMeasurement := strings.Split(c.config[i].Measurement, c.delimiter)
+		for j := range splitMeasurement {
+			if splitMeasurement[j] != "_" && splitMeasurement[j] != "" {
+				c.config[i].measurementIndex = j
+				break
+			}
+		}
+		c.config[i].splitTags = strings.Split(c.config[i].Tags, c.delimiter)
+		c.config[i].splitFields = strings.Split(c.config[i].Fields, c.delimiter)
+		c.config[i].splitSubject = strings.Split(c.config[i].Base, c.delimiter)
+
+		if len(splitMeasurement) != len(c.config[i].splitSubject) && len(splitMeasurement) != 1 {
+			err := fmt.Errorf("config error %s parsing: measurement length %d does not equal %s length %d",
+				c.baseName, len(splitMeasurement), c.baseName, len(c.config[i].splitSubject))
+			c.log.Error(err.Error())
```

Comment:
Please log the errors in the caller if necessary. They get returned anyway... Same for the spots below.